unit_scaling/parameter.py

```diff
@@ -0,0 +1,83 @@
+# Copyright (c) 2024 Graphcore Ltd. All rights reserved.
+
+"""Extends :class:`torch.nn.Parameter` with attributes for u-muP."""
+
+# mypy: disable-error-code="attr-defined, method-assign, no-untyped-call"
+
+from collections import OrderedDict
+from typing import Any, Dict, Literal, Optional, Protocol, TypeGuard
+
+import torch
+from torch import Tensor, nn
+
+MupType = Literal["weight", "bias", "norm", "output"]
+
+
+class ParameterData(Protocol):
+    """Extra fields for :class:`torch.nn.Parameter`, tagging u-muP metadata.
+
+    Objects supporting this protocol should implicitly also support
+    :class:`torch.nn.Parameter`.
+    """
+
+    mup_type: MupType
+    mup_scaling_depth: Optional[int]
+    shape: torch.Size  # repeated from nn.Parameter, for convenience
+
+
+def has_parameter_data(parameter: nn.Parameter) -> TypeGuard[ParameterData]:
+    """Check that the parameter supports the :class:`ParameterData` protocol."""
+    return (
+        getattr(parameter, "mup_type", None) in MupType.__args__
+        and hasattr(parameter, "mup_scaling_depth")
+        and isinstance(parameter.mup_scaling_depth, (type(None), int))
+    )
+
+
+def _parameter_deepcopy(self: nn.Parameter, memo: Dict[int, Any]) -> nn.Parameter:
```

Comment:
Can you test whether this `deepcopy` override is compatible with FakeTensor use in torch.compile?

Reply:
Thanks, yes, I've tried to write a test using `torch.compile`, but I'm not sure if it exercises the FakeTensor path you mention.